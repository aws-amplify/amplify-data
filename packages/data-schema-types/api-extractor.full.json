{
  "extends": "../../api-extractor.full.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/index.internal.d.ts"
}
