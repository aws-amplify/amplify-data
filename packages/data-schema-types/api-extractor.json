{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/index.d.ts"
}
