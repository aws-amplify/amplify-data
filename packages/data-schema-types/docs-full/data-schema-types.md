<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema-types](./data-schema-types.md)

## data-schema-types package

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[DerivedApiDefinition](./data-schema-types.derivedapidefinition.md)


</td><td>


</td></tr>
<tr><td>

[GraphQLFormattedError](./data-schema-types.graphqlformattederror.md)


</td><td>

See: https://spec.graphql.org/draft/\#sec-Errors


</td></tr>
<tr><td>

[SourceLocation](./data-schema-types.sourcelocation.md)


</td><td>

Represents a location in a Source.


</td></tr>
</tbody></table>

## Variables

<table><thead><tr><th>

Variable


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[\_\_modelMeta\_\_](./data-schema-types.__modelmeta__.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[AuthMode](./data-schema-types.authmode.md)


</td><td>


</td></tr>
<tr><td>

[BackendSecret](./data-schema-types.backendsecret.md)


</td><td>

Importing the full objects from @<!-- -->aws-amplify/plugin-types more than doubles dev env runtime. This type replacement will contain the content for config without the negative side-effects. We may need to re-approach if customers interact with these programmatically to avoid forcing narrowing.


</td></tr>
<tr><td>

[BooleanFilters](./data-schema-types.booleanfilters.md)


</td><td>


</td></tr>
<tr><td>

[ContextType](./data-schema-types.contexttype.md)


</td><td>


</td></tr>
<tr><td>

[CreateModelInput](./data-schema-types.createmodelinput.md)


</td><td>

All identifiers and fields used to create a model


</td></tr>
<tr><td>

[CustomHeaders](./data-schema-types.customheaders.md)


</td><td>

Custom headers that can be passed either to the client or to individual model operations, either as a static object or a function that returns a promise.


</td></tr>
<tr><td>

[CustomMutations](./data-schema-types.custommutations.md)


</td><td>


</td></tr>
<tr><td>

[CustomOperationFnParams](./data-schema-types.customoperationfnparams.md)


</td><td>

Generates Custom Operations function params based on whether .arguments() were specified in the schema builder


</td></tr>
<tr><td>

[CustomOperationMethodOptions](./data-schema-types.customoperationmethodoptions.md)


</td><td>


</td></tr>
<tr><td>

[CustomOperations](./data-schema-types.customoperations.md)


</td><td>


</td></tr>
<tr><td>

[CustomQueries](./data-schema-types.customqueries.md)


</td><td>


</td></tr>
<tr><td>

[CustomSelectionSetReturnValue](./data-schema-types.customselectionsetreturnvalue.md)


</td><td>

Generates flattened, readonly return type using specified custom sel. set


</td></tr>
<tr><td>

[CustomSqlDataSourceStrategy](./data-schema-types.customsqldatasourcestrategy.md)


</td><td>


</td></tr>
<tr><td>

[CustomSubscriptions](./data-schema-types.customsubscriptions.md)


</td><td>


</td></tr>
<tr><td>

[DataSourceConfiguration](./data-schema-types.datasourceconfiguration.md)


</td><td>


</td></tr>
<tr><td>

[DatasourceEngine](./data-schema-types.datasourceengine.md)


</td><td>


</td></tr>
<tr><td>

[Debug](./data-schema-types.debug.md)


</td><td>


</td></tr>
<tr><td>

[DeepPickFromPath](./data-schema-types.deeppickfrompath.md)


</td><td>

Picks object properties that match provided dot-separated Path


</td></tr>
<tr><td>

[DeepReadOnlyObject](./data-schema-types.deepreadonlyobject.md)


</td><td>


</td></tr>
<tr><td>

[DefineFunction](./data-schema-types.definefunction.md)


</td><td>


</td></tr>
<tr><td>

[DerivedCombinedSchema](./data-schema-types.derivedcombinedschema.md)


</td><td>


</td></tr>
<tr><td>

[DerivedModelSchema](./data-schema-types.derivedmodelschema.md)


</td><td>


</td></tr>
<tr><td>

[EnumTypes](./data-schema-types.enumtypes.md)


</td><td>

The utility type that is used to infer the type (interface) of the generated `client.enums` property.


</td></tr>
<tr><td>

[Equal](./data-schema-types.equal.md)


</td><td>


</td></tr>
<tr><td>

[ExcludeEmpty](./data-schema-types.excludeempty.md)


</td><td>


</td></tr>
<tr><td>

[Expect](./data-schema-types.expect.md)


</td><td>


</td></tr>
<tr><td>

[ExpectFalse](./data-schema-types.expectfalse.md)


</td><td>


</td></tr>
<tr><td>

[ExpectTrue](./data-schema-types.expecttrue.md)


</td><td>


</td></tr>
<tr><td>

[ExtractModelMeta](./data-schema-types.extractmodelmeta.md)


</td><td>


</td></tr>
<tr><td>

[FunctionSchemaAccess](./data-schema-types.functionschemaaccess.md)


</td><td>


</td></tr>
<tr><td>

[HasKey](./data-schema-types.haskey.md)


</td><td>


</td></tr>
<tr><td>

[IfEquals](./data-schema-types.ifequals.md)


</td><td>


</td></tr>
<tr><td>

[IndexQueryMethodsFromIR](./data-schema-types.indexquerymethodsfromir.md)


</td><td>


</td></tr>
<tr><td>

[IndexQueryMethodSignature](./data-schema-types.indexquerymethodsignature.md)


</td><td>


</td></tr>
<tr><td>

[IsAny](./data-schema-types.isany.md)


</td><td>


</td></tr>
<tr><td>

[IsEmptyStringOrNever](./data-schema-types.isemptystringornever.md)


</td><td>


</td></tr>
<tr><td>

[IsFalse](./data-schema-types.isfalse.md)


</td><td>


</td></tr>
<tr><td>

[IsTrue](./data-schema-types.istrue.md)


</td><td>


</td></tr>
<tr><td>

[JsResolver](./data-schema-types.jsresolver.md)


</td><td>


</td></tr>
<tr><td>

[JsResolverEntry](./data-schema-types.jsresolverentry.md)


</td><td>


</td></tr>
<tr><td>

[LambdaFunctionDefinition](./data-schema-types.lambdafunctiondefinition.md)


</td><td>


</td></tr>
<tr><td>

[LazyLoader](./data-schema-types.lazyloader.md)


</td><td>


</td></tr>
<tr><td>

[ListReturnValue](./data-schema-types.listreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[LogicalFilters](./data-schema-types.logicalfilters.md)


</td><td>


</td></tr>
<tr><td>

[Model](./data-schema-types.model.md)


</td><td>


</td></tr>
<tr><td>

[ModelFilter](./data-schema-types.modelfilter.md)


</td><td>


</td></tr>
<tr><td>

[ModelIdentifier](./data-schema-types.modelidentifier.md)


</td><td>


</td></tr>
<tr><td>

[ModelMetaShape](./data-schema-types.modelmetashape.md)


</td><td>


</td></tr>
<tr><td>

[ModelPath](./data-schema-types.modelpath.md)


</td><td>

Generates custom selection set type with up to 6 levels of nested fields


</td></tr>
<tr><td>

[ModelSortDirection](./data-schema-types.modelsortdirection.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypes](./data-schema-types.modeltypes.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesClient](./data-schema-types.modeltypesclient.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesSSRCookies](./data-schema-types.modeltypesssrcookies.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesSSRRequest](./data-schema-types.modeltypesssrrequest.md)


</td><td>


</td></tr>
<tr><td>

[MutationInput](./data-schema-types.mutationinput.md)


</td><td>

All required fields and relational fields, exclude readonly fields


</td></tr>
<tr><td>

[NonRelationalFields](./data-schema-types.nonrelationalfields.md)


</td><td>

Currently this omits any object-type fields. Update this when we add custom types/enums.


</td></tr>
<tr><td>

[NotAny](./data-schema-types.notany.md)


</td><td>


</td></tr>
<tr><td>

[NotEqual](./data-schema-types.notequal.md)


</td><td>


</td></tr>
<tr><td>

[NumericFilter](./data-schema-types.numericfilter.md)


</td><td>


</td></tr>
<tr><td>

[ObjectIsNonEmpty](./data-schema-types.objectisnonempty.md)


</td><td>


</td></tr>
<tr><td>

[ObservedReturnValue](./data-schema-types.observedreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[ObserveQueryReturnValue](./data-schema-types.observequeryreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[PathEntry](./data-schema-types.pathentry.md)


</td><td>

references IAmplifyGraphqlDefinition from: https://github.com/aws-amplify/amplify-category-api/blob/4c0ea253a0bae51f775383929ba4748593185bc1/packages/amplify-graphql-api-construct/src/types.ts\#L491-L503

function slots is any'd for now. Will add actual type when we add support for this feature


</td></tr>
<tr><td>

[Prettify](./data-schema-types.prettify.md)


</td><td>


</td></tr>
<tr><td>

[RequestOptions](./data-schema-types.requestoptions.md)


</td><td>

Request options that are passed to custom header functions. `method` and `headers` are not included in custom header functions passed to subscriptions.


</td></tr>
<tr><td>

[ResolvedModel](./data-schema-types.resolvedmodel.md)


</td><td>

Flattens model instance type and unwraps async functions into resolved GraphQL shape

This type is used for generating the base shape for custom selection set input and its return value Uses same pattern as above to limit recursion depth to maximum usable for selection set.


</td></tr>
<tr><td>

[RestoreArrays](./data-schema-types.restorearrays.md)


</td><td>

This mapped type traverses the SelectionSetReturnValue result and the original FlatModel, restoring array types that were flattened in DeepPickFromPath


</td></tr>
<tr><td>

[ReturnValue](./data-schema-types.returnvalue.md)


</td><td>

Selection set-aware CRUDL operation return value type


</td></tr>
<tr><td>

[SchemaConfiguration](./data-schema-types.schemaconfiguration.md)


</td><td>


</td></tr>
<tr><td>

[SecondaryIndexIrShape](./data-schema-types.secondaryindexirshape.md)


</td><td>

SecondaryIndex index types and query methods


</td></tr>
<tr><td>

[SelectionSet](./data-schema-types.selectionset.md)


</td><td>


</td></tr>
<tr><td>

[SetTypeSubArg](./data-schema-types.settypesubarg.md)


</td><td>

Replaces the value of a key in a complex generic type param


</td></tr>
<tr><td>

[SingularReturnValue](./data-schema-types.singularreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[SizeFilter](./data-schema-types.sizefilter.md)


</td><td>

Filter options that can be used on fields where size checks are supported.


</td></tr>
<tr><td>

[SqlStatementFolderEntry](./data-schema-types.sqlstatementfolderentry.md)


</td><td>


</td></tr>
<tr><td>

[StringFilter](./data-schema-types.stringfilter.md)


</td><td>

Filters options that can be used on string-like fields.


</td></tr>
<tr><td>

[SubnetAZ](./data-schema-types.subnetaz.md)


</td><td>


</td></tr>
<tr><td>

[UnionToIntersection](./data-schema-types.uniontointersection.md)


</td><td>


</td></tr>
<tr><td>

[UnwrapArray](./data-schema-types.unwraparray.md)


</td><td>


</td></tr>
<tr><td>

[VpcConfig](./data-schema-types.vpcconfig.md)


</td><td>


</td></tr>
<tr><td>

[WritableKeys](./data-schema-types.writablekeys.md)


</td><td>


</td></tr>
</tbody></table>
