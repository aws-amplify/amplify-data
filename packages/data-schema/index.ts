import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
