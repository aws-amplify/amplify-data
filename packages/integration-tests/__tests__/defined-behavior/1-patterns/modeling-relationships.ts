import { a, ClientSchema } from '@aws-amplify/data-schema';
import { Amplify } from 'aws-amplify';
import { type SelectionSet } from '@aws-amplify/data-schema-types';
import {
  buildAmplifyConfig,
  mockedGenerateClient,
  optionsAndHeaders,
  useState,
} from '../../utils';

describe('Modeling relationships', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Model one-to-many relationship', () => {
    // https://docs.amplify.aws/react/build-a-backend/data/data-modeling/relationships/#model-a-one-to-one-relationship
    const schema = a.schema({
      Team: a
        .model({
          mantra: a.string().required(),
          // 3. Create a hasMany relationship with the reference field
          //    from the `Member`s model.
          members: a.hasMany('Member', 'teamId'),
        })
        .authorization((allow) => [allow.publicApiKey()]),
      Member: a
        .model({
          name: a.string().required(),
          // 1. Create a reference field
          teamId: a.id(),
          // 2. Create a belongsTo relationship with the reference field
          team: a.belongsTo('Team', 'teamId'),
        })
        .authorization((allow) => [allow.publicApiKey()]),
    });
    type Schema = ClientSchema<typeof schema>;

    test('create a "Has Many" relationship between records', async () => {
      // #region mocking
      const expectedTeam = {
        __typeName: 'Team',
        id: 'some-team-id',
        mantra: 'Go Frontend!',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const expectedMember = {
        __typeName: 'Member',
        id: 'member-id',
        name: 'Tim',
        teamId: 'some-team-id',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            createTeam: expectedTeam,
          },
        },
        {
          data: {
            createMember: expectedMember,
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // #endregion mocking

      // #region docs code
      const { data: team } = await client.models.Team.create({
        mantra: 'Go Frontend!',
      });

      const { data: member } = await client.models.Member.create({
        name: 'Tim',
        teamId: team!.id,
      });
      // #endregion docs code

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(team).toEqual(expect.objectContaining(expectedTeam));
      expect(member).toEqual(expect.objectContaining(expectedMember));
      // #endregion assertions
    });

    test('Update a "Has Many" relationship between records', async () => {
      // #region mocking
      const expectedTeam = {
        __typeName: 'Team',
        id: 'new-team-id',
        mantra: 'Go Fullstack',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const expectedMember = {
        __typeName: 'Member',
        id: 'member-id',
        name: 'Tim',
        teamId: 'new-team-id',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            createTeam: expectedTeam,
          },
        },
        {
          data: {
            updateMember: expectedMember,
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // #endregion mocking

      // #region docs code
      const { data: newTeam } = await client.models.Team.create({
        mantra: 'Go Fullstack',
      });

      const { data: updatedMember } = await client.models.Member.update({
        id: 'member-id',
        teamId: newTeam!.id,
      });
      // #endregion docs code

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(newTeam).toEqual(expect.objectContaining(expectedTeam));
      expect(updatedMember).toEqual(expect.objectContaining(expectedMember));
      // #endregion assertions
    });

    test('Delete a "Has Many" relationship between records', async () => {
      // #region mocking
      const expectedMember = {
        __typeName: 'Member',
        id: 'member-id',
        name: 'Tim',
        teamId: null,
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            updateMember: expectedMember,
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // #endregion mocking

      // #region docs code
      const { data: member } = await client.models.Member.update({
        id: 'member-id',
        teamId: null,
      });
      // #endregion docs code

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(member).toEqual(expect.objectContaining(expectedMember));
      // #endregion assertions
    });

    test('Lazy load a "Has Many" relationship (hasMany side)', async () => {
      // #region mocking
      const expectedTeam = {
        __typeName: 'Team',
        id: 'new-team-id',
        mantra: 'Go Fullstack',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const expectedMember = {
        __typeName: 'Member',
        id: 'member-id',
        name: 'Tim',
        teamId: 'new-team-id',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            getTeam: expectedTeam,
          },
        },
        {
          data: {
            listMembers: [expectedMember],
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // #endregion mocking

      // #region App code
      const { data: team } = await client.models.Team.get({ id: 'team-id' });

      const { data: members } = await team!.members();

      members.forEach((member) => console.log(member.id));
      // #endregion App code

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(members[0]).toEqual(expect.objectContaining(expectedMember));
      // #endregion assertions
    });

    test('Lazy load a "Has Many" relationship (belongsTo side)', async () => {
      // Not in the docs.

      // #region mocking
      const expectedTeam = {
        __typeName: 'Team',
        id: 'new-team-id',
        mantra: 'Go Fullstack',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const expectedMember = {
        __typeName: 'Member',
        id: 'member-id',
        name: 'Tim',
        teamId: 'new-team-id',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            getMember: expectedMember,
          },
        },
        {
          data: {
            getTeam: expectedTeam,
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // #endregion mocking

      // #region App code
      const { data: member } = await client.models.Member.get({
        id: 'member-id',
      });

      const { data: team } = await member!.team();
      // #endregion App code

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(team).toEqual(expect.objectContaining(expectedTeam));
      // #endregion assertions
    });
  });

  describe('Model a "one-to-one" relationship', () => {
    // https://docs.amplify.aws/react/build-a-backend/data/data-modeling/relationships/#model-a-one-to-one-relationship
    const schema = a
      .schema({
        Cart: a.model({
          itemsX: a.string().required().array(),
          // 1. Create reference field
          customerId: a.id(),
          // 2. Create relationship field with the reference field
          customer: a.belongsTo('Customer', 'customerId'),
        }),
        Customer: a.model({
          name: a.string(),
          // 3. Create relationship field with the reference field
          //    from the Cart model
          activeCart: a.hasOne('Cart', 'customerId'),
        }),
      })
      .authorization((allow) => [allow.publicApiKey()]);
    type Schema = ClientSchema<typeof schema>;

    test('create a "Has One" relationship between records', async () => {
      // #region mocking
      const expectedCustomer = {
        __typeName: 'Customer',
        id: 'customer-id-rene',
        name: 'Rene',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const expectedCart = {
        __typeName: 'Cart',
        id: 'some-cart-id',
        customerId: 'customer-id-rene',
        itemsX: ['Tomato', 'Ice', 'Mint'],
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            createCustomer: expectedCustomer,
          },
        },
        {
          data: {
            createCart: expectedCart,
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // #endregion mocking

      // #region docs code
      const { data: customer, errors } = await client.models.Customer.create({
        name: 'Rene',
      });

      const { data: cart } = await client.models.Cart.create({
        itemsX: ['Tomato', 'Ice', 'Mint'],
        customerId: customer?.id,
      });
      // #endregion docs code

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(customer).toEqual(expect.objectContaining(expectedCustomer));
      expect(cart).toEqual(expect.objectContaining(expectedCart));
      // #endregion assertions
    });

    test('Update a "Has One" relationship between records', async () => {
      // #region mocking
      const expectedCustomer = {
        __typeName: 'Customer',
        id: 'customer-id-ian',
        name: 'Ian',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const expectedCart = {
        __typeName: 'Cart',
        id: 'some-cart-id',
        customerId: 'customer-id-ian',
        itemsX: ['Tomato', 'Ice', 'Mint'],
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            createCustomer: expectedCustomer,
          },
        },
        {
          data: {
            updateCart: expectedCart,
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // #endregion mocking

      // #region "state" from previous test
      const cart = {
        id: 'some-cart-id',
        customerId: 'some-customer-id',
        itemsX: ['Tomato', 'Ice', 'Mint'],
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };
      // #endregion "state" from previous test

      // #region docs code
      const { data: newCustomer } = await client.models.Customer.create({
        name: 'Ian',
      });

      const { data: updatedCart } = await client.models.Cart.update({
        id: cart.id,
        customerId: newCustomer?.id,
      });
      // #endregion docs code

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(newCustomer).toEqual(expect.objectContaining(expectedCustomer));
      expect(updatedCart).toEqual(expect.objectContaining(expectedCart));
      // #endregion assertions
    });

    test('Lazy load a "Has One" relationship (belongsTo side)', async () => {
      // #region mocking
      const expectedCustomer = {
        __typeName: 'Customer',
        id: 'customer-id',
        name: 'Leonard',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const expectedCart = {
        __typeName: 'Cart',
        id: 'some-cart-id',
        customerId: 'customer-id',
        itemsX: ['Tomato', 'Ice', 'Mint'],
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            getCart: expectedCart,
          },
        },
        {
          data: {
            getCustomer: expectedCustomer,
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // #endregion mocking

      // #region App code
      const { data: cart } = await client.models.Cart.get({
        id: 'some-cart-id',
      });
      const { data: customer } = await cart!.customer();
      // #endregion App code

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(customer).toEqual(expect.objectContaining(expectedCustomer));
      // #endregion assertions
    });

    test('Lazy load a "Has One" relationship (hasOne side)', async () => {
      // Not shown in the docs.

      // #region mocking
      const expectedCustomer = {
        __typeName: 'Customer',
        id: 'customer-id',
        name: 'Leonard',
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const expectedCart = {
        __typeName: 'Cart',
        id: 'some-cart-id',
        customerId: 'customer-id',
        itemsX: ['Tomato', 'Ice', 'Mint'],
        updatedAt: '2024-03-01T19:05:44.536Z',
        createdAt: '2024-03-01T18:05:44.536Z',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            getCustomer: expectedCustomer,
          },
        },
        {
          data: {
            // hasOne lazy loader performs a list and just selects the first item
            listCarts: [expectedCart],
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // #endregion mocking

      // #region App code
      const { data: retrievedCustomer } = await client.models.Customer.get({
        id: 'customer-id',
      });
      const { data: lazyLoadedCart } = await retrievedCustomer!.activeCart();
      // #endregion App code

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(lazyLoadedCart).toEqual(expect.objectContaining(expectedCart));
      // #endregion assertions
    });
  });
});
