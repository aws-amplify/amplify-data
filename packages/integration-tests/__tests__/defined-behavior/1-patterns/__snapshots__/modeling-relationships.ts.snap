// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modeling relationships Model a "one-to-one" relationship Lazy load a "Has One" relationship (belongsTo side) 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getCart(id: $id) { id itemsX customerId createdAt updatedAt } }",
      "variables": {
        "id": "some-cart-id",
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getCustomer(id: $id) { id name createdAt updatedAt } }",
      "variables": {
        "id": "customer-id",
      },
    },
    {},
  ],
]
`;

exports[`Modeling relationships Model a "one-to-one" relationship Lazy load a "Has One" relationship (hasOne side) 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getCustomer(id: $id) { id name createdAt updatedAt } }",
      "variables": {
        "id": "customer-id",
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelCartFilterInput,$limit: Int,$nextToken: String) { listCarts(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id itemsX customerId createdAt updatedAt } nextToken __typename } }",
      "variables": {
        "filter": {
          "and": [
            {
              "customerId": {
                "eq": "customer-id",
              },
            },
          ],
        },
      },
    },
    {},
  ],
]
`;

exports[`Modeling relationships Model a "one-to-one" relationship Update a "Has One" relationship between records 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreateCustomerInput!) { createCustomer(input: $input) { id name createdAt updatedAt } }",
      "variables": {
        "input": {
          "name": "Ian",
        },
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: UpdateCartInput!) { updateCart(input: $input) { id itemsX customerId createdAt updatedAt } }",
      "variables": {
        "input": {
          "customerId": "customer-id-ian",
          "id": "some-cart-id",
        },
      },
    },
    {},
  ],
]
`;

exports[`Modeling relationships Model a "one-to-one" relationship create a "Has One" relationship between records 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreateCustomerInput!) { createCustomer(input: $input) { id name createdAt updatedAt } }",
      "variables": {
        "input": {
          "name": "Rene",
        },
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreateCartInput!) { createCart(input: $input) { id itemsX customerId createdAt updatedAt } }",
      "variables": {
        "input": {
          "customerId": "customer-id-rene",
          "itemsX": [
            "Tomato",
            "Ice",
            "Mint",
          ],
        },
      },
    },
    {},
  ],
]
`;

exports[`Modeling relationships Model one-to-many relationship Delete a "Has Many" relationship between records 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: UpdateMemberInput!) { updateMember(input: $input) { id name teamId createdAt updatedAt } }",
      "variables": {
        "input": {
          "id": "member-id",
          "teamId": null,
        },
      },
    },
    {},
  ],
]
`;

exports[`Modeling relationships Model one-to-many relationship Eagerly load a "Has Many" relationship (hasMany side) 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getTeam(id: $id) { id members { items { id name teamId createdAt updatedAt } } } }",
      "variables": {
        "id": "team-id",
      },
    },
    {},
  ],
]
`;

exports[`Modeling relationships Model one-to-many relationship Lazy load a "Has Many" relationship (belongsTo side) 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getMember(id: $id) { id name teamId createdAt updatedAt } }",
      "variables": {
        "id": "member-id",
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getTeam(id: $id) { id mantra createdAt updatedAt } }",
      "variables": {
        "id": "new-team-id",
      },
    },
    {},
  ],
]
`;

exports[`Modeling relationships Model one-to-many relationship Lazy load a "Has Many" relationship (hasMany side) 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getTeam(id: $id) { id mantra createdAt updatedAt } }",
      "variables": {
        "id": "team-id",
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelMemberFilterInput,$limit: Int,$nextToken: String) { listMembers(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id name teamId createdAt updatedAt } nextToken __typename } }",
      "variables": {
        "filter": {
          "and": [
            {
              "teamId": {
                "eq": "new-team-id",
              },
            },
          ],
        },
      },
    },
    {},
  ],
]
`;

exports[`Modeling relationships Model one-to-many relationship Update a "Has Many" relationship between records 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreateTeamInput!) { createTeam(input: $input) { id mantra createdAt updatedAt } }",
      "variables": {
        "input": {
          "mantra": "Go Fullstack",
        },
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: UpdateMemberInput!) { updateMember(input: $input) { id name teamId createdAt updatedAt } }",
      "variables": {
        "input": {
          "id": "member-id",
          "teamId": "new-team-id",
        },
      },
    },
    {},
  ],
]
`;

exports[`Modeling relationships Model one-to-many relationship create a "Has Many" relationship between records 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreateTeamInput!) { createTeam(input: $input) { id mantra createdAt updatedAt } }",
      "variables": {
        "input": {
          "mantra": "Go Frontend!",
        },
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreateMemberInput!) { createMember(input: $input) { id name teamId createdAt updatedAt } }",
      "variables": {
        "input": {
          "name": "Tim",
          "teamId": "some-team-id",
        },
      },
    },
    {},
  ],
]
`;
